### PR TITLE
ci: upgrade builder image -> 4.5-1 (OTP 24.3.4.2-1)

### PR DESCRIPTION
**Pull request by thalesmg**

### CODE REVIEW

**Review by qzhuyan — APPROVED** [review]

**Comment by thalesmg** [comment]
Needs a release with #8696 , and probably hardcode some versions in the relup test specifically for 4.5.0.

File: build [hunk 1]
```diff
@@ -71,7 +71,14 @@ make_relup() {
     local lib_dir="_build/$PROFILE/rel/emqx/lib"
     local releases_dir="_build/$PROFILE/rel/emqx/releases"
     local name_pattern
-    name_pattern="${PROFILE}-$(./scripts/pkg-full-vsn.sh 'vsn_matcher')"
+    # 4.5.0 upgraded OTP from 24.1.5-3 to 24.3.4.2-1, so we need to
+    # fool the matcher to make it pick up the previous 4.4.X releases
+    # with older OTP.
+    if [[ "${PKG_VSN}" == 4.5.0* ]]; then
```
Comment (zmstone) [inline build:77]: we should consider the cases when upgrading from 4.4.[8-9] to 4.5.[0-9] ?

Comment (zmstone) [reply build:77]: for 4.5.* versions, we need to use this new pattern to match all 4.5 base versions.
but in addition, we should also list (fine to hard code) all possible base versions (4.4.8 onward)

Comment (thalesmg) [reply build:77]: The possible base versions are already enumerated in `relup-base-vsns.sh`.  I'll adapt that for more future 4.4+ and 4.5+ versions.

File: scripts/relup-base-packages.sh [hunk 2]
```diff
@@ -63,8 +63,21 @@ cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")/.."
...
+      ;;
+    4.5.*)
+      echo "24.3.4.2-1"
+      ;;
+  esac
+}
+
 for tag in $(../scripts/relup-base-vsns.sh $EDITION | xargs echo -n); do
-    filename="$PROFILE-${tag#[e|v]}-otp$OTP_VSN-$SYSTEM-$ARCH.zip"
+    filename="$PROFILE-${tag#[e|v]}-otp$(otp_vsn_for "$tag")-$SYSTEM-$ARCH.zip"
```
Comment (zmstone) [inline scripts/relup-base-packages.sh:80]: we no longer need line 31?

Comment (thalesmg) [reply scripts/relup-base-packages.sh:80]: Oops, better use the script for getting the OTP of 4.5+ than hard-coding the version.

File: build [hunk 3]
```diff
@@ -181,6 +188,11 @@ make_zip() {
             esac
             ;;
     esac
+    ## one can only upgrade to 4.5.X from 4.4.8 or later, when that's
+    ## released.  Until that's released, there's no relup.
+    if [[ -z "$(ls -A _upgrade_base 2>/dev/null || true)" && "${PKG_VSN}" == 4.5.0* ]]; then
+      has_relup='no'
+    fi
```
Comment (zmstone) [inline build:195]:
```suggestion
    ## one can only upgrade to 4.5.X from 4.4.8 or later, when that's
    ## released.  Until that's released, there's no relup.
    v44_bases="$(git tag -l "*4.4*" | grep -c -E '4\.4\.([8-9]|1[0-9])' || true)"
    if [[ "${v44_bases}" -eq 0 ]]; then
      has_relup='no'
    fi

File: scripts/relup-base-packages.sh (hunk 2 above)
Comment (zmstone) [reply scripts/relup-base-packages.sh:80]: I think it's fine to hard code here because we can not change the past.

**Comment by zmstone** [comment]
feel like there should be a place where we record all the versions etc for past releases.
e.g. `releases.json`
```
[{"version" : "4.4.8", "otp": "24.1.5-3", "hot_upgrade_from": ["4.4.0", "4.4.1", ...]},
  {"version": "4.5.0", "otp": "24.3..." 
```
this is committed in the repo, so we do not need to execute `git tag ... | grep` to fetch the versions.
also no need to hardcode opt versions etc.

File: scripts/relup-base-packages.sh (hunk 2 above)
Comment (thalesmg) [reply scripts/relup-base-packages.sh:80]: Sure, the old versions are indeed [hardcoded](https://github.com/emqx/emqx/pull/8684/files#diff-f42fb6f8358f71f02d16a191bb1e9a672e62a1a668fa2ef1cd9706806a758fe4R71), but I meant that the current one (for 4.5) could use the [current OTP](https://github.com/emqx/emqx/pull/8684/files#diff-f42fb6f8358f71f02d16a191bb1e9a672e62a1a668fa2ef1cd9706806a758fe4R74) from the script.  Does it make sense?

**Comment by thalesmg** [comment]
> 

Seems like a good idea to have something like that, but hardcoding the list of versions in a `hot_upgrade_from` fields may be error prone if someone forgets to add new versions to the list.

Maybe we could specify something like `hot_upgrade_from: ">= 4.4.8"` and use an Elixir script (in CI) to handle those [versions requirements](https://hexdocs.pm/elixir/Version.html), since it already has some facilities for that?

**Comment by thalesmg** [comment]
For example, something along the lines of:

```elixir
iex(9)> Version.match?("4.5.0", ">= 4.4.8")
true
iex(10)> Version.match?("4.5.1", ">= 4.4.8")
true
iex(11)> Version.match?("4.4.7", ">= 4.4.8")
false
iex(12)> Version.match?("4.5.1", ">= 4.4.8 and < 4.6.0")
true
iex(13)> Version.match?("4.6.1", ">= 4.4.8 and < 4.6.0")
false
```

**Comment by zmstone** [comment]
> For example, something along the lines of:
> 
> ```elixir
> iex(9)> Version.match?("4.5.0", ">= 4.4.8")
> true
> iex(10)> Version.match?("4.5.1", ">= 4.4.8")
> true
> iex(11)> Version.match?("4.4.7", ">= 4.4.8")
> false
> iex(12)> Version.match?("4.5.1", ">= 4.4.8 and < 4.6.0")
> true
> iex(13)> Version.match?("4.6.1", ">= 4.4.8 and < 4.6.0")
> false
> ```

or maybe this lib: https://github.com/jelly-beam/verl

File: build [hunk 4]
```diff
@@ -96,8 +103,8 @@ make_relup() {
     fi
     RELX_BASE_VERSIONS="$(IFS=, ; echo "${releases[*]}")"
     export RELX_BASE_VERSIONS
-    if [[ ${PKG_VSN} == 4.3* ]]; then
-        echo "EMQX 4.3 specific, overwrite OTP app versions"
+    if [[ ${PKG_VSN} == 4.3* || ${PKG_VSN} == 4.5.0* ]]; then
```
Comment (thalesmg) [inline build:106]: Maybe this needs to be `4.5.*`?

**Review by zmstone — APPROVED** [review]

**Pull request merged**